cmd/snap/cmd_info.go

```diff
@@ -0,0 +1,191 @@
+		return false
+	}
+	fmt.Fprintf(w, "path:\t%q\n", path)
+	fmt.Fprintf(w, "name:\t%s\n", info.Name())
+	fmt.Fprintf(w, "summary:\t%q\n", info.Summary())
+
+	var notes *Notes
+	if verbose {
+		fmt.Fprintln(w, "notes:\t")
+		fmt.Fprintf(w, "  confinement:\t%s\n", info.Confinement)
+		if info.Broken == "" {
+			fmt.Fprintln(w, "  broken:\tfalse")
+		} else {
+			fmt.Fprintf(w, "  broken:\ttrue (%s)\n", info.Broken)
+		}
+
+	} else {
+		notes = NotesFromInfo(info)
+	}
+	fmt.Fprintf(w, "version:\t%s %s\n", info.Version, notes)
+	if info.Type != snap.TypeApp {
+		fmt.Fprintf(w, "type:\t%s\n", info.Type)
+	}
+
+	return true
+}
+
+func coalesce(snaps ...*client.Snap) *client.Snap {
+	for _, s := range snaps {
+		if s != nil {
+			return s
+		}
+	}
+	return nil
+}
+
+func (x *infoCmd) Execute([]string) error {
+	cli := Client()
+
+	w := tabwriter.NewWriter(Stdout, 2, 2, 1, ' ', 0)
+
+	noneOK := true
+	for i, snapName := range x.Positional.Snaps {
+		if i > 0 {
+			fmt.Fprintln(w, "---")
+		}
+
+		if tryDirect(w, snapName, x.Verbose) {
+			noneOK = false
+			continue
+		}
+
+		remote, _, _ := cli.FindOne(snapName)
+		local, _, _ := cli.Snap(snapName)
+
+		both := coalesce(local, remote)
+
+		if both == nil {
+			fmt.Fprintf(w, "path:\t%q\nwarning:\t%s\n", snapName, i18n.G("not a valid snap"))
+			continue
+		}
+		noneOK = false
+
+		fmt.Fprintf(w, "name:\t%s\n", both.Name)
+		fmt.Fprintf(w, "summary:\t%q\n", both.Summary)
+		// TODO: have publisher; use publisher here,
+		// and additionally print developer if publisher != developer
+		fmt.Fprintf(w, "publisher:\t%s\n", both.Developer)
+		if x.Verbose {
+			fmt.Fprintln(w, "notes:\t")
+			fmt.Fprintf(w, "  private:\t%t\n", both.Private)
+			fmt.Fprintf(w, "  confinement:\t%s\n", both.Confinement)
+		}
+
+		if local != nil {
+			var notes *Notes
+			if x.Verbose {
+				jailMode := local.Confinement == client.DevModeConfinement && !local.DevMode
+				fmt.Fprintf(w, "  devmode:\t%t\n", local.DevMode)
+				fmt.Fprintf(w, "  jailmode:\t%t\n", jailMode)
+				fmt.Fprintf(w, "  trymode:\t%t\n", local.TryMode)
+				fmt.Fprintf(w, "  enabled:\t%t\n", local.Status == client.StatusActive)
+				if local.Broken == "" {
+					fmt.Fprintf(w, "  broken:\t%t\n", false)
+				} else {
+					fmt.Fprintf(w, "  broken:\t%t (%s)\n", true, local.Broken)
+				}
+			} else {
+				notes = NotesFromLocal(local)
+			}
+
+			fmt.Fprintf(w, "tracking:\t%s\n", local.Channel)
+			fmt.Fprintf(w, "installed:\t%s\t(%s)\t%s\n", local.Version, local.Revision, notes)
+		}
+
+		if remote != nil && remote.Channels != nil {
+			// \t\t\t so we get "installed" lined up with "channels"
+			fmt.Fprintf(w, "channels:\t\t\t\n")
+			for _, ch := range []string{"stable", "candidate", "beta", "edge"} {
+				m := remote.Channels[ch]
+				if m == nil {
+					continue
+				}
+				fmt.Fprintf(w, "  %s:\t%s\t(%s)\t%s\n", ch, m.Version, m.Revision, NotesFromRef(m))
+			}
+		}
+	}
+	w.Flush()
+
+	if noneOK {
+		return fmt.Errorf(i18n.G("no valid snaps given"))
```

Comment:
Agreed about the errors. I'll work on that in a separate branch (it promises to be tricky), as well as the de-coalescing (ditto). Meanwhile I'll ask Michael to add a "error handling of snap info still green" note to the release